runtime/fastly/builtins/fetch-event.cpp

```diff
@@ -0,0 +1,655 @@
+      cx, &JS::GetReservedSlot(self, static_cast<uint32_t>(Slots::Request)).toObject());
+
+  MOZ_ASSERT(!Request::request_handle(request).is_valid());
+
+  JS::SetReservedSlot(request, static_cast<uint32_t>(Request::Slots::Request),
+                      JS::Int32Value(req.handle));
+  JS::SetReservedSlot(request, static_cast<uint32_t>(Request::Slots::Body),
+                      JS::Int32Value(body.handle));
+
+  // Set the method.
+  auto res = req.get_method();
+  if (auto *err = res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  auto method_str = std::move(res.unwrap());
+  bool is_get = method_str == "GET"sv;
+  bool is_head = method_str == "HEAD"sv;
+
+  if (!is_get) {
+    JS::RootedString method(cx, JS_NewStringCopyN(cx, method_str.ptr.release(), method_str.len));
+    if (!method) {
+      return false;
+    }
+
+    JS::SetReservedSlot(request, static_cast<uint32_t>(Request::Slots::Method),
+                        JS::StringValue(method));
+  }
+
+  // Set whether we have a body depending on the method.
+  // TODO: verify if that's right. I.e. whether we should treat all requests
+  // that are not GET or HEAD as having a body, which might just be 0-length.
+  // It's not entirely clear what else we even could do here though.
+  if (!is_get && !is_head) {
+    JS::SetReservedSlot(request, static_cast<uint32_t>(Request::Slots::HasBody), JS::TrueValue());
+  }
+
+  auto uri_res = req.get_uri();
+  if (auto *err = uri_res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  auto uri_str = std::move(uri_res.unwrap());
+  JS::RootedString url(cx, JS_NewStringCopyN(cx, uri_str.ptr.get(), uri_str.len));
+  if (!url) {
+    return false;
+  }
+  JS::SetReservedSlot(request, static_cast<uint32_t>(Request::Slots::URL), JS::StringValue(url));
+
+  // Set the URL for `globalThis.location` to the client request's URL.
+  JS::RootedObject url_instance(cx, JS_NewObjectWithGivenProto(cx, &URL::class_, URL::proto_obj));
+  if (!url_instance) {
+    return false;
+  }
+
+  jsurl::SpecString spec(reinterpret_cast<uint8_t *>(uri_str.ptr.get()), uri_str.len, uri_str.len);
+  WorkerLocation::url = URL::create(cx, url_instance, spec);
+  if (!WorkerLocation::url) {
+    return false;
+  }
+
+  // Set `fastly.baseURL` to the origin of the client request's URL.
+  // Note that this only happens if baseURL hasn't already been set to another
+  // value explicitly.
+  if (!Fastly::baseURL.get()) {
+    JS::RootedObject url_instance(cx, JS_NewObjectWithGivenProto(cx, &URL::class_, URL::proto_obj));
+    if (!url_instance)
+      return false;
+
+    Fastly::baseURL = URL::create(cx, url_instance, URL::origin(cx, WorkerLocation::url));
+    if (!Fastly::baseURL)
+      return false;
+  }
+
+  return true;
+}
+
+// bool FetchEvent::init_incoming_request(JSContext *cx, JS::HandleObject self,
+//                                        host_api::HttpIncomingRequest *req) {
+//   JS::RootedObject request(
```

Comment:
Is this in-progress, or can we remove the commented-out code?

Reply:
I'ce gone ahead and removed this.